libvmi/driver/xen/xen.c

```diff
@@ -2455,15 +2455,51 @@ xen_set_vcpureg(
 #if defined(ARM32) || defined(ARM64)
     return xen_set_vcpureg_arm(vmi, value, reg, vcpu);
 #elif defined(I386) || defined (X86_64)
-    if (!xen_get_instance(vmi)->hvm) {
-        if (8 == xen_get_instance(vmi)->addr_width) {
-            return xen_set_vcpureg_pv64(vmi, value, reg, vcpu);
-        } else {
-            return xen_set_vcpureg_pv32(vmi, value, reg, vcpu);
+    if (xen_get_instance(vmi)->hvm) {
+        switch (reg) {
+        case RAX:
+        case RBX:
+        case RCX:
+        case RDX:
+        case RBP:
+        case RSI:
+        case RDI:
+        case RSP:
+        case R8:
+        case R9:
+        case R10:
+        case R11:
+        case R12:
+        case R13:
+        case R14:
+        case R15:
+        case RIP:
+        case RFLAGS:
+        case CR0:
+        case CR2:
+        case CR3:
```

Comment:
Unfortunately if the guest is using PCID and we set CR3 through here we would have some unintended consequences (PCID uses the lower 12 bits of the CR3 as the TLB tag). So IMHO that cr3 sanitization should only take place if the guest is PV.